scheduling-utils/src/handshake/server.rs

```diff
@@ -0,0 +1,279 @@
+                    })
+                },
+            )?;
+        let (worker_to_pack_files, worker_to_pack) = (0..logon.worker_count)
+            .map(|_| Self::create_producer(logon.worker_to_pack_size))
+            .try_fold(
+                (Vec::default(), Vec::default()),
+                |(mut files, mut queues), res| {
+                    res.map(|(file, queue)| {
+                        files.push(file);
+                        queues.push(queue);
+
+                        (files, queues)
+                    })
+                },
+            )?;
+
+        // Send the file descriptors to the client.
+        let fds: Vec<_> = [&tpu_to_pack_file, &progress_tracker_file]
+            .into_iter()
+            .chain(&pack_to_worker_files)
+            .chain(&worker_to_pack_files)
+            .map(|file| file.as_raw_fd())
+            .collect();
+        let iov = [IoSlice::new(&[LOGON_SUCCESS])];
+        let cmsgs = [ControlMessage::ScmRights(&fds)];
+        let sent =
+            socket::sendmsg::<UnixAddr>(stream.as_raw_fd(), &iov, &cmsgs, MsgFlags::empty(), None)
+                .map_err(std::io::Error::from)?;
+        debug_assert_eq!(sent, 1);
+
+        Ok(Session {
+            tpu_to_pack,
+            progress_tracker,
+            pack_to_worker,
+            worker_to_pack,
+        })
+    }
+
+    fn create_producer<T>(size: usize) -> Result<(File, shaq::Producer<T>), ShaqError> {
+        let file = Self::create_shmem(size)?;
+        let queue = shaq::Producer::create(&file, size)?;
+
+        Ok((file, queue))
+    }
+
+    fn create_consumer(
+        size: usize,
+    ) -> Result<(File, shaq::Consumer<PackToWorkerMessage>), ShaqError> {
+        let file = Self::create_shmem(size)?;
+        let queue = shaq::Consumer::create(&file, size)?;
+
+        Ok((file, queue))
+    }
+
+    fn create_shmem(size: usize) -> Result<File, ShaqError> {
+        let size_t = i64::try_from(size).map_err(|_| ShaqError::InvalidBufferSize)?;
+
+        // Create a memory backed file.
+        #[cfg(any(
+            target_os = "linux",
+            target_os = "l4re",
+            target_os = "android",
+            target_os = "emscripten"
+        ))]
+        let file = Self::memfd_create()?;
+        #[cfg(not(any(
+            target_os = "linux",
+            target_os = "l4re",
+            target_os = "android",
+            target_os = "emscripten"
+        )))]
+        let file = Self::shmem_create()?;
+
+        // Resize to the requested size.
+        let ret = unsafe { libc::ftruncate64(file.as_raw_fd(), size_t) };
+        if ret == -1 {
+            return Err(ShaqError::Io(std::io::Error::last_os_error()));
+        }
+        debug_assert_eq!(ret, 0);
+
+        Ok(file)
+    }
+
+    #[cfg(any(
+        target_os = "linux",
+        target_os = "l4re",
+        target_os = "android",
+        target_os = "emscripten"
+    ))]
+    fn memfd_create() -> Result<File, std::io::Error> {
```

Comment:
note to self - we'll want to support hugepages here, but we'll need to detect and use approporiate flags in shaq/rts-alloc as well. we also only want to use for some of the obj I think (not the progress message queue I think, since it's so small)